runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,162 @@
+local M = {}
+
+local semantic_tokens = setmetatable({}, {
+  __index = function(table, key)
+    local rtn = {}
+    rawset(table, key, rtn)
+    return rtn
+  end,
+})
+local last_tick = {}
+local active_requests = {}
+
+---@private
+local function get_bit(n, k)
+  --todo(theHamsta): remove once `bit` module is available for non-LuaJIT
+  if _G.bit then
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  else
+    return (n / math.pow(2, k)) % 2
+  end
+end
+
+---@private
+local function modifiers_from_number(x, modifiers_table)
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+--- 
+--- This function can be configured with |vim.lsp.with()| with the following options for `config`
+--- 
+--- `on_token`: A function with signature `function(ctx, token)` that is called
+---             whenever a semantic token is received from the server from context `ctx`
+---             (see |lsp-handler| for the definition of `ctx`). This can be used for highlighting the tokens.
+---             `token` is a table:
+---
+--- <pre>
+---   {
+---         line       -- line number 0-based
+---         start_char -- start character 0-based
+---         length     -- length in characters of this token
+---         type       -- token type as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         modifiers  -- token modifier as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens 
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  if not ctx.bufnr or not ctx.client_id then
+    return
+  end
+  active_requests[ctx.bufnr] = false
```

Comment:
This is a good illustration of what the current client.request API is suboptimal. Ideally we should just cancel the request via the request ID when a user has taken an action that would invalidate the semantic tokens.

This is generally a problem whenever applying highlights from the language server... In reality, if the *only* changes from the last update tick are on the current line, we should be able to apply highlights everywhere else in the document (but the current line)....

@mfussenegger one option might be to expose lastline/new_lastline ranges to the handlers, like "these were the changed ranges since the request was fired"

Reply:
>  reality, if the only changes from the last update tick are on the current line, we should be able to apply highlights everywhere else in the document (but the current line)....


Not necessarily. For example if the current line contains a declaration of a variable, that could impact all references of the variable as well.

Reply:
I think we can keep this implemention for now. It's the same as for the code lense implemention and can be refactored together in a follow-up

Reply:
> Not necessarily. For example if the current line contains a declaration of a variable, that could impact all references of the variable as well.

The current state of this PR will just invalidate all highlights if the buffer has changed since sending the request. Even if the edit does impact the declaration of the variable, that would change subsequent highlights.